[ENH] Add SplineTrendForecaster

jgyasu opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs

Fixes: #7363



#### What does this implement/fix? Explain your changes.

Implemented SplineTrendForecaster based on the BaseForecaster, according to the feature request, I have made a similar forecaster as the PolynomialTrendForecaster but SplineTransformer from sklearn is used as the regressor


#### Does your contribution introduce a new dependency? If yes, which one?

No

#### What should a reviewer concentrate their feedback on?

~I have chosen `knots="quantile` and `extrapolation="periodic"` as the default parameter while initialising the SplineTransformer, I request a feedback on it~  I made a new commit with the parameters which are default in scikit-learn based on the suggestion in the friday meet

#### Did you add any tests for the change?

No, I did not add any tests for the change, I used `check_estimator` and it passed all the tests

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators


<!--
Thanks for contributing!
-->

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,215 @@
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Implements SplineTrendForecaster."""
+
+__author__ = ["tensorflow-as-tf", "mloning", "aiwalter", "fkiraly", "jgyasu"]
```
I think this may be just you, if this class is yours entirely. If you made this way to note that it was inspired by polynomial trend one, I believe you can consider inheriting or similar if there is a significant amount of code overlap.

(Just my opinoin, not necessarily of all core developers)

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,215 @@
+
+
+class SplineTrendForecaster(BaseForecaster):
+    """Forecast time series data with a spline trend."""
```
Can you please update this docstring with `__init__` arguments, examples, references, etc.?

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,215 @@
+        "capability:pred_int": True,
+    }
+
+    def __init__(
```
Will it make sense to expose `n_knots` from [`SplineTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.SplineTransformer.html) as well?

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,215 @@
+        """
+        from sklearn.ensemble import RandomForestRegressor
+
+        params_list = [
+            {},
+            {
+                "regressor": RandomForestRegressor(),
+                "degree": 2,
+                "with_intercept": False,
+                "prediction_intervals": False,
+            },
+            {
+                "regressor": RandomForestRegressor(),
+                "degree": 2,
+                "with_intercept": True,
+                "prediction_intervals": True,
+            },
+            {
+                "regressor": RandomForestRegressor(),
+                "degree": 2,
+                "with_intercept": False,
+                "prediction_intervals": True,
+            },
+        ]
```
Without increasing the number of test paramas, can you cover more init arguments like extrapolation or knots please?

yarnabrina reviewed (changes requested):
Hi @jgyasu , thank you for the contribution. I've made some comments, can you kindly take a look?

Also, please add your new estimator in `docs/source/api_reference/forecasting.rst`.

jgyasu commented:
Hi @yarnabrina, thank you very much for the feedback. I have considered the changes and updated the code and made a new commit.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+        - 'periodic': Uses periodic splines with a periodicity equal to the distance
+        between the first and last knot, enforcing equal function values and
+        derivatives at these knots.
+    include_bias : bool, default=True
```
This is missing from your `__init__`.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+    """
+
+    _tags = {
+        "authors": ["jgyasu", "tensorflow-as-tf", "mloning", "aiwalter", "fkiraly"],
```
I suggest to make it just you. The dependency has been acknowledged through classs inheritance IMO, @fkiraly can comment further.

```suggestion
        "authors": ["jgyasu"],
```

However, you should consider adding a maintainer tag too.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+        # intervals is not calculated. If set to True, the extra information is
+        # calculated and stored in the forecaster.
+
+        self.set_tags(**{"capability:pred_int": prediction_intervals})
```
Question: is this required? I believe it's handled in parent class already.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    def __init__(
+        self,
+        regressor=None,
+        degree=1,
+        with_intercept=True,
+        prediction_intervals=False,
+        n_knots=5,
+        knots="uniform",
+        extrapolation="constant",
```
I suggest a reorder of arguments, followed by similar changes in docstring. My suggestion is first argument being regressor, followed by all arguments you plan to expose from SplineTransformer, followed by arguments necessary by parent class that are not covered yet. Let me know what you think.

jgyasu replied:
Thanks, that's indeed a better organisation, I have reordered the arguments.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+                n_knots=self.n_knots,
+                knots=self.knots,
+                extrapolation=self.extrapolation,
+                include_bias=False,
```
Why are you forcing `False`? Should you not use intercept or bias arguments?

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+            fitted_values = self.regressor_.predict(X_sklearn)
+            residuals = y - fitted_values
+            p = self.degree + int(self.with_intercept)
+            self.s_squared_ = np.sum(residuals**2) / (len(y) - p)
+            self.train_index_ = y.index
```
Question to @fkiraly @benHeid : I remember there were some discussions about remembering data and/or not, do not remember details. If we store this type of info for history, do we need to check some config (store only if allowed or etc.)?

fkiraly replied:
Obsolete due to removal of prediction intervals.

More generally, this is fine - the estimator itself can decide what to remember.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+
+        return self
+
+    def _predict(self, fh=None, X=None):
```
Can you please tell the difference with polynomial? Why is that not enough?

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+        y_pred.name = self._y.name
+        return y_pred
+
+    def _predict_var(self, fh=None, X=None, cov=False):
```
Same as above, is it not identical to that of `PolynomialTrendForecaster`?

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+                "prediction_intervals": True,
+                "n_knots": 5,
+                "knots": "uniform",
+                "extrapolation": "constant",
```
Please change to other values, don't fix to just same value. Same for degree, n_knots, knots.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,291 @@
+
+
+class SplineTrendForecaster(PolynomialTrendForecaster):
+    """
```
I think it's preferred to start docstring on same line as initial triple quotes. See this: https://numpydoc.readthedocs.io/en/latest/validation.html#built-in-validation-checks

yarnabrina reviewed (changes requested):
Made few more comments. Also, please update with `main` branch post the latest release.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,223 @@
+        self.n_knots = n_knots
+        self.knots = knots
+        self.extrapolation = extrapolation
+        # prediction_intervals : bool, default=False
+        # By default, the extra information needed to later generate the prediction
+        # intervals is not calculated. If set to True, the extra information is
+        # calculated and stored in the forecaster.
```
Optional: Not sure what is this comment for, probably can be removed.

yarnabrina commented on sktime/forecasting/trend/_spline_trend_forecaster.py:
```diff
@@ -0,0 +1,223 @@
+
+    References
+    ----------
+    .. [1] Hyndman, Rob J., and George Athanasopoulos. Forecasting: principles
+    and practice, 3rd edition. OTexts: Melbourne, Australia. OTexts.com/fpp3.
```
Optional: While you should give original book reference, I think to help users you may give direct link to relevant section at https://otexts.com/fpp3/regression-matrices.html as well.

yarnabrina reviewed (dismissed):
Looks good to me. Made two minor optional comments, and the previous question to @benHeid / @fkiraly remains open.

fkiraly reviewed (dismissed):
This should get merged with https://github.com/sktime/sktime/pull/7502 which uses the delegation approach.

fkiraly commented:
Some notes:

* I think the prediction interval calculations are only valid for the polynomial forecaster - I removed it hence
* I moved the base class to be `BaseForecaster` to avoid unnecessarily complex inheritance chains.

jgyasu commented:
> Some notes:
> 
> * I think the prediction interval calculations are only valid for the polynomial forecaster - I removed it hence
> * I moved the base class to be `BaseForecaster` to avoid unnecessarily complex inheritance chains.

I see, we have decided to go with the `BaseForecaster` now. But I have been reading documentation and I understand why `_DelegatedForecaster` was supposed to be used. However, due to a lack of proper documentation on `_DelegatedForecaster`, I couldn't progress much with it. It'd be really helpful if you can maybe go over `_DelegatedForecaster` a bit in the next community meetings @fkiraly

fkiraly commented:
Sure!

I will try to merge with https://github.com/sktime/sktime/pull/7502 before the next release, and see whether it is easy to introduce the `_DelegatedForecaster` via the merge.
